### ISSUE TITLE
The player sometimes recieves damage after death and respawning.

### ISSUE THREAD

**Issue by orbea** [minetest/minetest#9738]
##### Minetest version
```
$ minetest --version
Minetest 5.2.0 (Linux)
Using Irrlicht 1.8.4
BUILD_TYPE=Release
RUN_IN_PLACE=0
USE_GETTEXT=1
USE_SOUND=1
USE_CURL=1
USE_FREETYPE=1
USE_LUAJIT=1
STATIC_SHAREDIR="/usr/share/minetes
```
multidimensions: https://github.com/AiTechEye/multidimensions/commit/aaf75c49abf8539f35adf208fe9319a9666dff94

##### OS / Hardware
```
Operating system: Slackware64 Current
CPU: AMD FX-6350
```

##### Summary

This seems to be a continuation or related to issue https://github.com/minetest/minetest/issues/3565.

Basically when testing the multidimensions mod and then teleporting to the floatlands from the mod I found that when falling the player will hit a defined node that does 9000 `damage_per_second` which likely instantly kills the player. However if the player quickly presses the respawn button they will respawn and then die one more time. This does not happen if they wait several seconds before respawning.

I tried other ways of killing the player and was able to reproduce damage after death with lava at least once, but its not at all consistent.

Multidimensions registers the killing node here.
```
minetest.register_node("multidimensions:killing", {
	description = "Killing",
	drawtype="airlike",
	groups = {unbreakable=1,not_in_creative_inventory = 1},
	paramtype = "light",
	sunlight_propagates = true,
	drop = "",
	walkable=false,
	damage_per_second = 9000,
	pointable=false,
	diggable = false,
})
```
https://github.com/AiTechEye/multidimensions/blob/aaf75c49abf8539f35adf208fe9319a9666dff94/api.lua#L291

And places at in floatlands.
```
multidimensions.register_dimension("floatandlike",{
...
	self={
		blocking="multidimensions:blocking",
		killing = "multidimensions:killing",
	},
	on_generate=function(self,data,id,area,x,y,z)
		if y <= self.dirt_start-70 then
			data[id] = self.killing
```
https://github.com/AiTechEye/multidimensions/blob/aaf75c49abf8539f35adf208fe9319a9666dff94/dimensions.lua#L71

I was able to work around this issue by placing an early `return` in `Server::handleCommand_Damage()` before the `playersao->setHP()` line. This is obviously not a solution.

https://github.com/minetest/minetest/blob/914dbeaa0be4b5ef87506b605ef4e241cd3732dc/src/network/serverpackethandler.cpp#L819

From stdout it shows the player dies twice.
```
2020-04-24 07:32:21: ACTION[Server]: singleplayer [127.0.0.1] joins game. List of players: singleplayer
2020-04-24 07:32:28: ACTION[Server]: singleplayer dies at (-63,4387,76). No bones placed
2020-04-24 07:32:29: ACTION[Server]: singleplayer respawns at (-63,4516,65)
2020-04-24 07:32:29: ACTION[Server]: singleplayer damaged by 70 hp at (-63,4516,65)
2020-04-24 07:32:29: ACTION[Server]: singleplayer dies at (-63,4516,65). No bones placed
2020-04-24 07:32:30: ACTION[Server]: singleplayer respawns at (-63,4516,65)
2020-04-24 07:32:39: ACTION[Main]: Server: Shutting down
```

##### Steps to reproduce

1. Install the multidimensions mod. https://github.com/AiTechEye/multidimensions
2. Grab the `multidimensions:teleporter_floatlandlike` node.
3. Place it and right click it to teleport to floatlands.
4. Find the nearest ledge and jump off.
5. When dying press respawn as fast as possible.

As a side note if the `damage_per_second` is disabled the player will fall and fall and fall and then hit land without dying...?

**Comment by orbea** [comment]
I think there is some kind of race condition with damage and respawning? Maybe it should set damage to 0 when respawning the first time to avoid this problem?

**Comment by An0n3m0us** [comment]
Confirmed:
https://my.pcloud.com/publink/show?code=XZlDswkZQpbOnrxu0H4lrhIjcxR9CHbOzDny

Other odd behavior occurs too, as shown on the third teleport.

**Comment by ClobberXD** [comment]
Can confirm this bug. For me, it's the persisting velocity that's the devil. If I somehow die mid-fall, I die again when I respawn.

This always happens to me in ITB, wherein I'd be teleported to the start of a box but the mapblocks wouldn't have emerged yet. I start falling immediately after teleporting, and insta-die once I reach y = -50 (ITB is coded that way to prevent players from falling to the bottom of the bottomless world). I consistently respawn around 3-4 times due to this bug.

**Comment by Wuzzy2** [comment]
Does this mod mess with the player armor? Because if it erases the `immortal` group, thinks break.

**Comment by orbea** [comment]
> Does this mod mess with the player armor? Because if it erases the immortal group, thinks break.

Not that I can tell, no results with my grep at least.

```
$ grep -ri immortal multidimensions/
```

**Comment by orbea** [comment]
I think I have a basic idea what is happening.

* The player dies and triggers the death screen.

https://github.com/minetest/minetest/blob/bc60e44d80a02bd48163440da2fd3165c0910376/src/client/game.cpp#L2601

* The player's speed remains constant since nothing stops the player's fall.

* When respawning the player hits the ground again and dies due to the high speed before dying the first time.

https://github.com/minetest/minetest/blob/bc60e44d80a02bd48163440da2fd3165c0910376/src/client/clientenvironment.cpp#L283

This is easy to workaround by setting the `multidimensions:killing` node to `walkable=true`, but I'm struggling how to make sure the player's speed is set to 0 in the engine after respawning.

**Comment by Wuzzy2** [comment]
I can confirm orbea's claims. Bug confirmed in 5.3.0 and 4f2303849e0f929524695a9e3719ec486b47ddd1.

*Here's a simplified and reliable list of steps to reproduce:*

1. Grant yourself all privs
2. Make sure the area from Y=0 to Y=200 is loaded (at your current X/Z position)
3. Teleport to Y=200
4. Disable fly mode to fall down
5. Use the `/kill` command while falling
6. Press "Respawn" BEFORE you hit the floor

Expected outcome: You respawn safely
Actual outcome: You respawn, but keep the speed that your corpse had before respawning, so you will likely take fall damage right after respawn. The fall damage might be fatal, killing you again

My analysis:
* The multidimensions mod is irrelevant
* When you die while falling, and then respawn (while your corpse is still falling), your speed is preserved

So basically, what orbea said.

So I guess to fix this, Minetest needs to reset the speed to 0 after respawn.

**Comment by Zughy** [comment]
I've just tried to set player speed to zero via `playersao->getPlayer()->setSpeed(v3f(0,0,0));`, both on death AND on respawn, but it still happens :thinking:

**Comment by birdlover32767** [comment]
maybe you need to do it on the clientside too? seems like you did it only on the serverside
